password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,243 @@
+    -- ^ Required password minimum length
+    , passPolicyMaximumLength :: !Int
+    -- ^ Required password maximum length
+    , passPolicyCharUppercase :: !(Maybe Int)
+    -- ^ Required number of upper-case characters
+    , passPolicyCharLowercase :: !(Maybe Int)
+    -- ^ Required number of lower-case characters
+    , passPolicyCharSpecial   :: !(Maybe Int)
+    -- ^ Required number of special characters
+    , passPolicyCharDigit     :: !(Maybe Int)
+    -- ^ Required number of ASCII-digit characters
+    , passPolicyCharSet       :: !Text
+    -- ^ Set of characters that can be used for password
+    }
+    deriving (Eq, Ord, Show)
+
+-- | Generate valid PasswordPolicy
+instance Arbitrary PasswordPolicy where
+  arbitrary = do
+    minimumLength <- choose (1, 10)
+    upperCase <- genMaybeInt
+    lowerCase <- genMaybeInt
+    special <- genMaybeInt
+    digit <- genMaybeInt
+    let sumLength = sum $ catMaybes [upperCase, lowerCase, special, digit]
+    let minMaxLength = maximum [minimumLength, sumLength]
+    maximumLength <- choose (minMaxLength, minMaxLength + 10)
+    return $ PasswordPolicy minimumLength maximumLength upperCase lowerCase special digit defaultCharSet
+    where
+      genMaybeInt :: Gen (Maybe Int)
+      genMaybeInt = oneof [return Nothing, Just <$> (choose (1, 10))]
+
+-- | Default value for the 'PasswordPolicy'
+defaultPasswordPolicy :: PasswordPolicy
+defaultPasswordPolicy = PasswordPolicy
+  { passPolicyMinimumLength = 8,
+    passPolicyMaximumLength = 32,
+    passPolicyCharUppercase = Just 1,
+    passPolicyCharLowercase = Just 1,
+    passPolicyCharSpecial = Nothing,
+    passPolicyCharDigit = Just 1,
+    passPolicyCharSet = defaultCharSet
+  }
+
+-- | Default character sets consist of uppercase, lowercase letters, numbers,
+-- and special characters
+defaultCharSet :: Text
+defaultCharSet = T.pack $ specialChars <> ['A' .. 'Z'] <> ['a' .. 'z'] <> ['0' .. '9']
+
+-- | Special characters
+specialChars :: String
+specialChars = " !\"#$%&'()*+,-./:;<=>?@[]^_`{|}~"
+
+-- | Character Category
+data CharacterCategory = Uppercase
+    | Lowercase
+    | Special
+    | Digit
+    deriving (Eq, Ord, Show)
+
+instance Arbitrary CharacterCategory where
+  arbitrary = elements [Uppercase, Lowercase, Special, Digit]
+
+-- | Data type representing how password are invalid
+data InvalidReason = PasswordTooShort !Int !Int
+    | PasswordTooLong !Int !Int
+    | NotEnoughReqChars CharacterCategory !Int !Int
+    | InvalidChar !Text
+    | InvalidPasswordPolicy !PasswordPolicy
+    deriving (Eq, Ord, Show)
+
+-- | Check if given 'Password' fullfills all the Policies,
+-- return true if given password is valid
+--
+-- This is equivalent to @null $ validatePassword policy password@
+--
+-- >>> let pass = mkPassword "This_Is_Valid_PassWord1234"
+-- >>> valid defaultPasswordPolicy pass
+-- True
+valid :: PasswordPolicy -> Password -> Bool
+valid policy pass = null $ validatePassword policy pass
```

Comment:
`valid` is a bit unnecessarily short. `isValidPassword` would be preferred, IMHO.
`isValidPassword policy = null . validatePassword policy`